// typedoc/dist/lib/converter/plugins/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypePlugin = exports.SourcePlugin = exports.PackagePlugin = exports.LinkResolverPlugin = exports.InheritDocPlugin = exports.ImplementsPlugin = exports.GroupPlugin = exports.CommentPlugin = exports.CategoryPlugin = void 0;
var CategoryPlugin_1 = require("./CategoryPlugin");
Object.defineProperty(exports, "CategoryPlugin", { enumerable: true, get: function () { return CategoryPlugin_1.CategoryPlugin; } });
var CommentPlugin_1 = require("./CommentPlugin");
Object.defineProperty(exports, "CommentPlugin", { enumerable: true, get: function () { return CommentPlugin_1.CommentPlugin; } });
var GroupPlugin_1 = require("./GroupPlugin");
Object.defineProperty(exports, "GroupPlugin", { enumerable: true, get: function () { return GroupPlugin_1.GroupPlugin; } });
var ImplementsPlugin_1 = require("./ImplementsPlugin");
Object.defineProperty(exports, "ImplementsPlugin", { enumerable: true, get: function () { return ImplementsPlugin_1.ImplementsPlugin; } });
var InheritDocPlugin_1 = require("./InheritDocPlugin");
Object.defineProperty(exports, "InheritDocPlugin", { enumerable: true, get: function () { return InheritDocPlugin_1.InheritDocPlugin; } });
var LinkResolverPlugin_1 = require("./LinkResolverPlugin");
Object.defineProperty(exports, "LinkResolverPlugin", { enumerable: true, get: function () { return LinkResolverPlugin_1.LinkResolverPlugin; } });
var PackagePlugin_1 = require("./PackagePlugin");
Object.defineProperty(exports, "PackagePlugin", { enumerable: true, get: function () { return PackagePlugin_1.PackagePlugin; } });
var SourcePlugin_1 = require("./SourcePlugin");
Object.defineProperty(exports, "SourcePlugin", { enumerable: true, get: function () { return SourcePlugin_1.SourcePlugin; } });
var TypePlugin_1 = require("./TypePlugin");
Object.defineProperty(exports, "TypePlugin", { enumerable: true, get: function () { return TypePlugin_1.TypePlugin; } });
